lib/AST/DiagnosticEngine.cpp

```diff
@@ -1518,7 +1512,7 @@ DiagnosticEngine::diagnosticInfoForDiagnostic(const Diagnostic &diagnostic) {
 
   return DiagnosticInfo(
       diagnostic.getID(), loc, toDiagnosticKind(behavior),
-      diagnosticStringFor(diagnostic.getID(), getPrintDiagnosticNames()),
+      diagnosticStringFor(diagnostic.getID(), getPrintDiagnosticNamesMode()),
       diagnostic.getArgs(), Category, getDefaultDiagnosticLoc(),
```

Comment:
It looks like `Category` here has a few hardcoded values. Should we provide the warning group name for `Category` when there is one?

Reply:
I think we probably could do it. But I'm sure we should. These categories are inferred from DiagnosticOptions.
The `deprecation` category seems to be interesting to look at. Maybe we should consider removing the `DiagnosticOptions::Deprecation` option and instead infer this category from checking if a diagnostic belongs to the `deprecated` group. So we don't have two ways to express the same thing.
And I don't know if we need the `api-digester-breaking-change` and `no-usage` categories expressed as groups.

Reply:
I took a look at the diagnostics that get each of these flags: the "deprecation" category does seem to line up fairly well with the "deprecated" group, the "no-usage" category could be useful as a general diagnostic group for "you did something and didn't use the result" warnings, and "api-digester-breaking-change" is a very specific group of separate diagnostics (mostly errors).